src/dist/component/components.rs

```diff
@@ -145,15 +154,38 @@ impl<'a> ComponentBuilder<'a> {
 }
 
 #[derive(Debug)]
-pub struct ComponentPart(pub String, pub PathBuf);
+pub struct ComponentPart {
+    /// Kind of the [`ComponentPart`], such as `"file"` or `"dir"`.
+    pub kind: String,
```

Comment:
It's pre-existing but could this be an enum?

Reply:
@ChrisDenton Thanks! I've made [`b4323b7` (#4179)](https://github.com/rust-lang/rustup/pull/4179/commits/b4323b7e9e93f949984e895d43d483f3496012a3).

Meanwhile, I chose to keep an `Unknown` variant to maintain bidirectional conversions. I also didn't use `kind: Result<ComponentPartKind, String>` because I don't think we are really using any `Result`-related facilities in that particular case. A hard `.parse()` error would look weird as well, since the original logic handles unknown `ComponentPartKind`s much later in the business logic.